### PR TITLE
Change slot_type verification strategy

### PR DESCRIPTION
**Pull request by NBadyal**
We ran into an issue where `get_slot_type` kept returning `None` on primitive types and strings in several configurations (`sequence<float>`, `string<25>`, `sequence<string>`, etc.). Also saw that static arrays of items were not being validated in the gui, i.e. a `sequence<string<25>, 5>` could have strings with 26+ characters without indicating an error.

In an attempt to minimize the `get_slot_type` issue, I used regex matching to define the type based on the user's input. I also added `array` functionality to the `_eval_locals`.

When evaluating a static array, instead of inspecting the individual item, I remove the index in the `topic_name`, copy the message instance's version of the array, insert the expression, and then set the expression to the array. This allows the array to be validated.

### CODE REVIEW

File: src/rqt_publisher/publisher.py [hunk 1]
```diff
@@ -227,39 +227,59 @@ def _change_publisher_expression(self, publisher_info, topic_name, new_value):
...
-            success, _ = self._evaluate_expression(expression, slot_type)
+                user_expression = user_expression[:error_prefix_pos]
+
+            computed_expression = str(user_expression)
+            # reduce indexed topic_names, if necessary
+            topic_name_reduction = re.match("(.*)\[[0-9]*\]$", topic_name)
+            if topic_name_reduction:
+                topic_name = topic_name_reduction.group(1)
+
+            # sequences
```
Comment (gbiggs) [inline src/rqt_publisher/publisher.py:244]: Please add a comment describing what this code block is doing and why.

File: src/rqt_publisher/publisher.py [hunk 2]
```diff
@@ -227,39 +227,59 @@ def _change_publisher_expression(self, publisher_info, topic_name, new_value):
             # Strip topic name from the full topic path
             slot_path = topic_name.replace(publisher_info['topic_name'], '', 1)
             slot_path, slot_array_index = self._extract_array_info(slot_path)
+            slot_type = None
 
-            # Get the property type from the message class
-            slot_type, is_array = \
-                get_slot_type(publisher_info['message_instance'].__class__, slot_path)
```
Comment (gbiggs) [inline src/rqt_publisher/publisher.py:233]: Where did the `get_slot_type()` function call go and why is it no longer necessary? Have you removed the function if it's no longer being used?

**Review by gbiggs — COMMENTED** [review]
Please retarget your PR to the `rolling` branch.

It would be good to have a test that shows the behaviour you're trying to correct, then shows that your change corrects it.

**Comment by NBadyal** [comment]
@gbiggs First, thank you for the feedback. 

For testing, I created a custom message with uint8, floats, and string variants and entered in the same data for each, in the same order. `rolling` has multiple issues ranging from not being able to evaluate `array('x', [...])`(i.e. `name 'array' is not defined`), determining slot_type (i.e. `AttributeError: 'NoneType'`), failing to evaluate expressions derived from `numpy`, etc. The PR changes produce no errors on the same input.

rqt_test_input

On `rolling`:
```bash
RMW Implementation: rmw_apex_middleware
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-nicholas.badyal'
[ERROR] [get_message_class]: Malformed message_type: sequence<float>
Python eval failed for expression "array('f', [1.2, 3.4, 5.6])" with an exception "name 'array' is not defined"
Publisher._evaluate_expression(): failed to evaluate expression: "array('f', [1.2, 3.4, 5.6])" as Python type "None"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 262, in _change_publisher_expression
    expression, error_prefix, slot_type.__name__)
AttributeError: 'NoneType' object has no attribute '__name__'
[ERROR] [get_message_class]: Malformed message_type: sequence<float, 5>
Python eval failed for expression "array('f', [5.4, 3.2])" with an exception "name 'array' is not defined"
Publisher._evaluate_expression(): failed to evaluate expression: "array('f', [5.4, 3.2])" as Python type "None"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 262, in _change_publisher_expression
    expression, error_prefix, slot_type.__name__)
AttributeError: 'NoneType' object has no attribute '__name__'
[ERROR] [get_message_class]: Malformed message_type: sequence<uint8>
Python eval failed for expression "array('B', [1, 2, 3, 4, 5])" with an exception "name 'array' is not defined"
Publisher._evaluate_expression(): failed to evaluate expression: "array('B', [1, 2, 3, 4, 5])" as Python type "None"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 262, in _change_publisher_expression
    expression, error_prefix, slot_type.__name__)
AttributeError: 'NoneType' object has no attribute '__name__'
[ERROR] [get_message_class]: Malformed message_type: sequence<uint8, 5>
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Python eval failed for expression "array('B', [5, 4, 3, 2, 1])" with an exception "name 'array' is not defined"
Publisher._evaluate_expression(): failed to evaluate expression: "array('B', [5, 4, 3, 2, 1])" as Python type "None"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 262, in _change_publisher_expression
    expression, error_prefix, slot_type.__name__)
AttributeError: 'NoneType' object has no attribute '__name__'
[ERROR] [get_message_class]: Malformed message_type: string<25>
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: sequence<string>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: sequence<string<25>>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: string<25>
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "1" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "2" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "3" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "4" as Python type "<class 'numpy.uint8'>"
Publisher._evaluate_expression(): failed to evaluate expression: "5" as Python type "<class 'numpy.uint8'>"
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: string<25>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: string<25>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: string<25>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: string<25>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: sequence<string, 5>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
[ERROR] [get_message_class]: Malformed message_type: sequence<string<25>, 5>
Traceback (most recent call last):
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 161, in change_publisher
    new_text = handler(self._publishers[publisher_id], topic_name, new_value)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 244, in _change_publisher_expression
    success, _ = self._evaluate_expression(expression, slot_type)
  File "/home/nicholas.badyal/data/workspace/build/rqt_publisher/src/rqt_publisher/publisher.py", line 329, in _evaluate_expression
    if successful_eval and isinstance(value, slot_type):
TypeError: isinstance() arg 2 must be a type or tuple of types
```

PR Changes:
```bash
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-nicholas.badyal'
RMW Implementation: rmw_apex_middleware
```

File: src/rqt_publisher/publisher.py [hunk 3]
```diff
@@ -227,39 +227,59 @@ def _change_publisher_expression(self, publisher_info, topic_name, new_value):
...
-            if slot_array_index is not None:
-                is_array = False
-
-            if is_array:
-                slot_type = list
             # strip possible trailing error message from expression
-            error_prefix = '# error'
-            error_prefix_pos = expression.find(error_prefix)
+            error_prefix = ' # error'
```
[the remark this answers is not in the document]
Comment (NBadyal) [reply src/rqt_publisher/publisher.py]: Replaced with:
```python
# strip possible trailing error message from expression
contains_error = re.match("(.*)\s*# error.*", user_expression)
if contains_error:
    user_expression = contains_error.group(1)
```

File: src/rqt_publisher/publisher.py [hunk 4]
```diff
@@ -227,39 +227,59 @@ def _change_publisher_expression(self, publisher_info, topic_name, new_value):
...
+            error_prefix_pos = user_expression.find(error_prefix)
             if error_prefix_pos >= 0:
-                expression = expression[:error_prefix_pos]
-            success, _ = self._evaluate_expression(expression, slot_type)
+                user_expression = user_expression[:error_prefix_pos]
+
+            computed_expression = str(user_expression)
+            # reduce indexed topic_names, if necessary
```
[the remark this answers is not in the document]
Comment (NBadyal) [reply src/rqt_publisher/publisher.py]: Changed to:
```python
# expression can contain topics with indexes, i.e. /chatter[0]
# remove index to match message_instance, i.e. /chatter
topic_name_includes_index = re.match("(.*)\[[0-9]*\]$", topic_name)
if topic_name_includes_index:
    topic_name = topic_name_includes_index.group(1)
```

File: src/rqt_publisher/publisher.py (hunk 1 above)
Comment (NBadyal) [reply src/rqt_publisher/publisher.py]: Changed to:
```python
# static sequences change one item at a time, impossible to validate
# handle as entire sequence, enables validation
if slot_array_index is not None:
```

File: src/rqt_publisher/publisher.py (hunk 2 above)
Comment (NBadyal) [reply src/rqt_publisher/publisher.py:233]: Added comment to clarify where it's replaced:
```python
# determine the property type, supplemental wrapper for get_slot_type()
slot_type = \
      self._resolve_slot_type(computed_expression, slot_type, slot_path, publisher_info['message_instance'].__class__)
```

Added comments in function to clarify why it's still used:
```python
def _resolve_slot_type(self, expression, slot_type, slot_path, message_class):
        if slot_type is None:
            # check for types that get_slot_type doesn't handle well
            # check for array.array type
            is_array_array = re.search("^array\(\'.\', (\[.*\])\)", expression)
            if is_array_array:
                return array.array

            # check for list type
            is_list = re.match("^\[.*\]\s*$", expression)
            if is_list:
                return list

            # check for string type
            is_string = re.match("^\'.*\'\s*$", expression)
            if is_string:
                return str

            # at this point, get_slot_type can determine the type
            slot_type, _ = \
                get_slot_type(message_class, slot_path)

        return slot_type
```

**Comment by gbiggs** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17004/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11565/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17414/)

**Comment by gbiggs** [comment]
CI looks good so going ahead and merging this. Thanks for the contribution!

**Pull request merged**